File: ios/ClaudeScope/Sources/Services/AnalyticsService.swift
import Foundation
#if canImport(AppsFlyerLib)
import AppsFlyerLib
#endif

/// Centralized analytics tracking for AppsFlyer events.
@MainActor
final class AnalyticsService {
    static let shared = AnalyticsService()

    // MARK: - Configuration

    /// Call once at app launch to initialize AppsFlyer SDK.
    func configure() {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().appsFlyerDevKey = "F8FrZFW2PUgiJG3bXT9Y2E"
        AppsFlyerLib.shared().appleAppID = "6761888911"
        // Wait up to 60s for ATT prompt resolution before firing the install event.
        // iOS 14.5+ hides IDFA until the user answers ATT; without this delay the
        // install attribution would miss IDFA on first launch.
        AppsFlyerLib.shared().waitForATTUserAuthorization(timeoutInterval: 60)
        #if DEBUG
        AppsFlyerLib.shared().isDebug = true
        #else
        AppsFlyerLib.shared().isDebug = false
        #endif
        #endif
    }

    /// Call when app becomes active (triggers AppsFlyer session).
    func start() {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().start()
        #endif
    }

    // MARK: - Auth Events

    func trackLogin(method: String = "oauth") {
        #if canImport(AppsFlyerLib)
        // Fire both events: AFEventLogin marks the interactive sign-in action;
        // AFEventCompleteRegistration marks the first successful auth handshake
        // (Meta and AppsFlyer use CompleteRegistration for ad attribution ROAS).
        AppsFlyerLib.shared().logEvent(AFEventLogin, withValues: [
            AFEventParamRegistrationMethod: method,
        ])
        AppsFlyerLib.shared().logEvent(AFEventCompleteRegistration, withValues: [
            AFEventParamRegistrationMethod: "claude_oauth",
        ])
        #endif
    }

    func trackSignOut() {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("sign_out", withValues: [:])
        #endif
    }

    // MARK: - Usage Events

    func trackUsageFetched(pct5h: Double, pct7d: Double) {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("usage_fetched", withValues: [
            "pct_5h": Int(round(pct5h * 100)),
            "pct_7d": Int(round(pct7d * 100))
        ])
        #endif
    }

    func trackThresholdAlert(window: String, pct: Int) {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("threshold_alert", withValues: [
            "window": window,
            "pct": pct
        ])
        #endif
    }

    // MARK: - Navigation Events

    func trackTabSelected(_ tab: String) {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("tab_selected", withValues: [
            "tab": tab
        ])
        #endif
    }

    func trackLanguageChanged(_ language: String) {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("language_changed", withValues: [
            "language": language
        ])
        #endif
    }

    func trackPollingIntervalChanged(_ minutes: Int) {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("polling_interval_changed", withValues: [
            "minutes": minutes
        ])
        #endif
    }

    // MARK: - Widget Events

    func trackWidgetConfigured(size: String) {
        #if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent("widget_configured", withValues: [
            "size": size
        ])
        #endif
    }
}

File: android/app/src/main/kotlin/io/sandwichlab/claudescope/service/analytics/AnalyticsService.kt
package io.sandwichlab.claudescope.service.analytics

import android.app.Application
import android.util.Log
import com.appsflyer.AFInAppEventParameterName
import com.appsflyer.AFInAppEventType
import com.appsflyer.AppsFlyerLib
import com.facebook.FacebookSdk
import com.facebook.appevents.AppEventsLogger
import io.sandwichlab.claudescope.BuildConfig
import kotlin.math.roundToInt

/**
 * Centralised analytics tracking for AppsFlyer + Facebook SDK events.
 *
 * Event surface is a direct port of iOS [AnalyticsService.swift] — same names,
 * same param keys, same value formats — so the two platforms roll up into the
 * same AppsFlyer project without munging.
 *
 * Usage:
 * ```
 *   // Once, from Application.onCreate:
 *   AnalyticsService.configure(this)
 *   AnalyticsService.start(this)
 *
 *   // Anywhere:
 *   AnalyticsService.trackTabSelected("home")
 * ```
 */
object AnalyticsService {

    private const val TAG = "AnalyticsService"
    private const val DEV_KEY = BuildConfig.APPSFLYER_DEV_KEY

    @Volatile private var configured = false

    /** Initialise SDKs. Safe to call multiple times — subsequent calls no-op. */
    fun configure(application: Application) {
        if (configured) return
        configured = true

        runCatching {
            AppsFlyerLib.getInstance().apply {
                setDebugLog(BuildConfig.DEBUG)
                init(DEV_KEY, null, application)
            }
        }.onFailure { Log.w(TAG, "AppsFlyer init failed", it) }

        runCatching {
            FacebookSdk.sdkInitialize(application)
            AppEventsLogger.activateApp(application)
        }.onFailure { Log.w(TAG, "Facebook SDK init failed", it) }
    }

    /** Start AppsFlyer session. Call when the app becomes interactive. */
    fun start(application: Application) {
        runCatching {
            AppsFlyerLib.getInstance().start(application)
        }.onFailure { Log.w(TAG, "AppsFlyer start failed", it) }
    }

    // MARK: - Auth events

    fun trackLogin(method: String = "oauth") = logEvent(
        AFInAppEventType.LOGIN,
        mapOf(AFInAppEventParameterName.REGISTRATION_METHOD to method),
    )

    fun trackSignOut() = logEvent("sign_out", emptyMap())

    // MARK: - Usage events

    fun trackUsageFetched(pct5h: Double, pct7d: Double) = logEvent(
        "usage_fetched",
        mapOf(
            "pct_5h" to (pct5h * 100).roundToInt(),
            "pct_7d" to (pct7d * 100).roundToInt(),
        ),
    )

    fun trackThresholdAlert(window: String, pct: Int) = logEvent(
        "threshold_alert",
        mapOf("window" to window, "pct" to pct),
    )

    // MARK: - Navigation events

    fun trackTabSelected(tab: String) = logEvent("tab_selected", mapOf("tab" to tab))

    fun trackLanguageChanged(language: String) =
        logEvent("language_changed", mapOf("language" to language))

    fun trackPollingIntervalChanged(minutes: Int) =
        logEvent("polling_interval_changed", mapOf("minutes" to minutes))

    // MARK: - Widget events

    fun trackWidgetConfigured(size: String) =
        logEvent("widget_configured", mapOf("size" to size))

    // MARK: - Internal

    private fun logEvent(event: String, params: Map<String, Any>) {
        if (!configured) return
        runCatching {
            AppsFlyerLib.getInstance().logEvent(null, event, params)
        }.onFailure { Log.w(TAG, "AppsFlyer logEvent '$event' failed", it) }
    }
}
